service/frontend/task_reachability.go

```diff
@@ -48,38 +53,36 @@ import (
 type versionSetFetcher struct {
 	lock           sync.Mutex
 	matchingClient matchingservice.MatchingServiceClient
-	futures        map[string]future.Future[*matchingservice.GetWorkerBuildIdCompatibilityResponse]
+	futures        map[string]future.Future[*persistencespb.VersioningData]
 }
 
 func newVersionSetFetcher(matchingClient matchingservice.MatchingServiceClient) *versionSetFetcher {
 	return &versionSetFetcher{
 		matchingClient: matchingClient,
-		futures:        make(map[string]future.Future[*matchingservice.GetWorkerBuildIdCompatibilityResponse]),
+		futures:        make(map[string]future.Future[*persistencespb.VersioningData]),
 	}
 }
 
-func (f *versionSetFetcher) getFuture(ctx context.Context, ns *namespace.Namespace, taskQueue string) future.Future[*matchingservice.GetWorkerBuildIdCompatibilityResponse] {
+func (f *versionSetFetcher) getFuture(ctx context.Context, ns *namespace.Namespace, taskQueue string) future.Future[*persistencespb.VersioningData] {
 	f.lock.Lock()
 	defer f.lock.Unlock()
 	_, found := f.futures[taskQueue]
 	if !found {
-		fut := future.NewFuture[*matchingservice.GetWorkerBuildIdCompatibilityResponse]()
+		fut := future.NewFuture[*persistencespb.VersioningData]()
 		f.futures[taskQueue] = fut
 		go func() {
-			value, err := f.matchingClient.GetWorkerBuildIdCompatibility(ctx, &matchingservice.GetWorkerBuildIdCompatibilityRequest{
-				NamespaceId: ns.ID().String(),
-				Request: &workflowservice.GetWorkerBuildIdCompatibilityRequest{
-					Namespace: ns.Name().String(),
-					TaskQueue: taskQueue,
-				},
+			value, err := f.matchingClient.GetTaskQueueUserData(ctx, &matchingservice.GetTaskQueueUserDataRequest{
```

Comment:
do you think it makes sense to use the more general GetTaskQueueUserData in frontend's GetWorkerBuildIdCompatibility also and remove that specific rpc from matching? it's not really doing anything besides ToBuildIdOrderingResponse, which frontend could call.

Reply:
Sure, that makes sense to me.

TBH, I would just send all of the user facing user data info in DescribeTaskQueue and not have a special API for GetWorkerBuildIdCompatibility.

Reply:
In any case, I wouldn't change in this PR.
We'd need to leave the matching RPC for this minor version because 1.21 was already released.

Reply:
you mean not have GetWorkerBuildIdCompatibility in workflowservice (frontend)? I like that but we released it already, can we remove it? for matchingservice we need a separate GetTaskQueueUserData for long polls anyway so might as well use that

Reply:
We marked this whole feature as experimental (API-wise), we can still change but we'll likely need to deprecate APIs and keep them around for a while since users will start relying on them fairly soon.